File: 2015/03/solution.py
data = [l.strip() for l in open(0).read()]
part1 = set()
santa = (0,0)

def step(pos,dir):
    r,c = pos
    if dir == '^':
        r-=1
    elif dir == 'v':
        r+=1
    elif dir == '<':
        c -= 1
    elif dir == '>':
        c += 1
    else:
        raise("Wrong step")
    return (r,c)    

for s in data:
    santa = step(santa,s)    
    part1.add(santa)
    
print(len(part1))

chunked_steps = [data[2*i:2*i+2] for i in range(0,len(data)//2)]
part2 = set()
santa = (0,0)
robo = (0,0)
for s in chunked_steps:
    sdir, rdir = s
    santa = step(santa,sdir)    
    robo = step(robo,rdir)
    part2.add(santa)
    part2.add(robo)

print(len(part2))

File: 2015/03/solution.rb
# frozen_string_literal: true

input = $stdin.read.strip.chars
pos = 0 + 0.i

def step(pos, dir)
  case dir
  when '^'
    pos - 1.i
  when 'v'
    pos + 1.i
  when '<'
    pos - 1
  when '>'
    pos + 1
  end
end

santa_path = input.each_with_object([pos]) do |step, path|
  curr_pos = path[-1]
  next_pos = step(curr_pos, step)
  path << next_pos
end
p santa_path.to_set.size
part2 = input.each_slice(2).each_with_object([[0 + 0.i], [0 + 0.i]]) do |steps, paths|
  santa_path, robot_path = paths
  santa_step, robot_step = steps
  santa_path << step(santa_path[-1], santa_step)
  robot_path << step(robot_path[-1], robot_step)
  [santa_path, robot_path]
end
p part2.flatten.to_set.size
